### DISCUSSION TITLE
State timeline panel: Pagination or scroll view support

### DISCUSSION THREAD

**Discussion by ashleyjlive** [grafana/grafana#38303] [Dashboards, Panels & Visualizations]
**What would you like to be added**:
Add pagination **or** scroll view support to state timeline panel.

**Why is this needed**:
State timeline has the potential to be a very powerful and valuable tool for monitoring uptime of a large number of nodes. In my initial testing I found that for a high number rows the view tends to squish and fit all rows into the fixed view - which makes it very hard if not impossible to read [1]. It would be better to either:
• Have pagination (page numbers, etc)
• Allow overflow and provide a scroll view within the window rather than trying to fit everything in the window

My preference is for a scroll view as that fits more with the style of what the graph and time series does when their legends contain more data than what can fit within the window.

[1]
![image](https://user-images.githubusercontent.com/9400511/122309256-5331b800-cf51-11eb-8417-b74c7e5dc302.png)

**Comment by trollvottel** [comment]
+1

Without this, the classic 3rdparty 'discrete' panel is still the preference. I'd even call this a bug, because usually all panels may scroll internally.

**Comment by trollvottel** [comment]
![grafik](https://user-images.githubusercontent.com/9528457/122409568-868d3a80-cf83-11eb-9e3d-22d32156a9ab.png)

Looks quite useless, doesn't it?

**Comment by ryantxu** [comment]
In the discrete panel -- we configured a row height and then that scrolled within the panel.  We could consider something like that

![Peek 2021-06-17 10-29](https://user-images.githubusercontent.com/705951/122445957-f507d380-cf56-11eb-8faf-0914a3555e3f.gif)

**Comment by ashleyjlive** [comment]
> +1
> 
> Without this, the classic 3rdparty 'discrete' panel is still the preference. I'd even call this a bug, because usually all panels may scroll internally.

Yeah agreed. I wouldn't really call this a feature request.

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Reply by trollvottel** [reply]
6 months later, still not scrolling in v8.3.4. What a shame. Well triaged, pkolyvas, well triaged.

**Comment by cmawhorter** [comment]
because of this, this visualization is pretty much only useful for a static data source.  in the alarm screenshot above, or in my case (instances in an ASG), it's difficult to know the right height to set ahead of time since it's a moving target.

in fact, this is a universal issue for all panels that display dynamic data.  the legend can end up overtaking the graph.

**Comment by BenGrenoble** [comment]
Hello,
This feature also interest me. I'm coming here to +1 this bug.

**Comment by wdlaszlo** [comment]
+1 for this.

**Comment by DmitrySidorow** [comment]
1 year and this basic feauture still not implemented, discussions sucks

**Reply by trollvottel** [reply]
True. Sadly I'm not typescript-capable enough to fix this myself.

**Comment by nickfaughey** [comment]
This must be a bug right? All other panels scroll vertically by default. Makes this panel unusable if you have either 1) many different series to visualize, or 2) repeated panels (as in rows) with _different_ numbers of series (you can make this panel super tall to account for the lack of scrolling, but if the next row only has 2 series, it just looks dumb being that tall...)

Can this be prioritized differently as a bug? The classic Discrete panel plugin doesn't even cover our use case because it doesn't display nulls properly like the State Timeline does.

**Reply by leeoniya** [reply]
> This must be a bug right?

yes, it's a bug. so probably should go into issues.

> All other panels scroll vertically by default.

do they? to my knowledge only "list" style panels do, like Table, Dashboard list, Alert list, News, Annotation list. what all these panels have in common is they are rendered in HTML. in all other panels the viz is sized to the panel content area and does not scroll. i dont't know which non-list panels are currently sized based on their contents rather than the contents being scaled down to fit the panel dimensions.

so this issue is tricky to fix because:

1. it needs the canvas/viz size to be governed by the data, not panel height. which is not a thing we currently do, afaik.
2. it needs to have an extra option to fixate the height of each row if you want to have a uniform look across different datasets.
3. it needs to have its x axis pinned to the bottom of the scroll container, which means clipping part of the tall canvas and creating an additional canvas to which we copy the rendered x axis. is it possible to pin something with the scrollbars library we use without destroying performance? maybe, idk :shrug:
4. things like tooltips and context menus may need to be adjusted to ensure they account for scroll offsets within the viz content area.
5. Status history panel will also need to have this implemented, since it is similar.
6. probably other stuff that isn't coming to mind right now

horizontal Bar chart may conceivably also want this.

i will try to get around to prototyping all this in the next few months, but it is not a trivial fix, unfortunately.

**Reply by nickfaughey** [reply]
Appreciate the insight and the triage! I'm not too familiar with the inner architecture of the viz renderings, so I didn't realize this was a fundamentally hard problem.

My squished State Timeline is sitting right next to a few horizontal Bar Gauges that scroll beautifully vertically, and I imagined that the State Timeline would end up rendered very similar to the Bar Gauge but with rich x-axis data... alas it seems they're two very different visualization frameworks that just happen to look structurally similar.

**Reply by leeoniya** [reply]
> I imagined that the State Timeline would end up rendered very similar to the Bar Gauge

a bar guage requires 1 dom element per bar (unless it's the retro/LCD style one :sweat:, i have plans to fix this!). so if you have 100 bars, that's 100 DOM nodes -- not a huge deal. on the other hand, state timeline could have _thousands_ of individual rects _per lane_, since the state can change every 100ms to a different color or threshold. so if you have 100 lanes, you could have a _hundred thousand_ DOM nodes, which would be catastrophic, so this viz is implemented on canvas. fundamentally they are very different.

**Reply by nickfaughey** [reply]
Yes this makes perfect sense, and I appreciate the explanation.

**Reply by cmawhorter** [reply]
@leeoniya have you experimented with svg instead of canvas?  it's still nodes, but probably more performant than html nodes and could handle the scroll/tooltips natively.

**Reply by leeoniya** [reply]
svg sucks the same way :(

**Comment by ghost** [comment]
We really need this - scrolling, with fixed height, both as options, would make this usable.

Right now, we can't really use it.

**Comment by akhilesh0** [comment]
+1

**Comment by woutkonings** [comment]
+1

**Comment by BogdanIonesq** [comment]
Not sure why the issue was closed, as without this, the State Timeline is pretty much not usable. Moreover, with Grafana v10.x, the Discrete plugin is no longer supported.

**Comment by gelicia** [comment]
Hello! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

This discussion was converted into an issue, which is available [here](https://github.com/grafana/grafana/issues/73436).
